### ISSUE TITLE
[Bug Report] TGS Velocity Discrepancy

### ISSUE THREAD

**Issue by ori-gadot** [isaac-sim/IsaacLab#1493]
### Bug Description
The observed joint velocity is incorrect.

### Steps to Reproduce

1. Use the provided minimal single-file script to reproduce the issue.
2. The script runs the CartPole environment with joint position as actions and prints the joint position, joint velocity, and the numerical derivative of the position.
3. The CartPole moves from its initial pose to the target pose and remains stationary.
4. Despite reaching a static state, the reported joint velocity is a positive and relatively large value.

### Additional Context

Switching the solver type from TGS (1) to PGS (0) in the following line shows a notable difference in accuracy:
`env_cfg.sim.physx.solver_type = 0`

Using PGS yields more accurate velocity reporting in the static state.

### Acceptance Criteria
When the system is in a static state, the reported joint velocity should be as close to zero as possible.

### Code

```
from omni.isaac.lab.app import AppLauncher

app_launcher = AppLauncher(headless=True)
simulation_app = app_launcher.app

import torch
import omni.isaac.lab.envs.mdp as mdp
from omni.isaac.lab.envs import ManagerBasedEnv, ManagerBasedEnvCfg
from omni.isaac.lab.managers import ObservationGroupCfg as ObsGroup
from omni.isaac.lab.managers import ObservationTermCfg as ObsTerm
from omni.isaac.lab.utils import configclass
from omni.isaac.lab_tasks.manager_based.classic.cartpole.cartpole_env_cfg import CartpoleSceneCfg
@configclass
class ActionsCfg:
    joint_pos = mdp.JointPositionActionCfg(asset_name="robot", joint_names=["cart_to_pole"], scale=1.0, use_default_offset=False)

@configclass
class ObservationsCfg:
    @configclass
    class PolicyCfg(ObsGroup):
        joint_pos_rel = ObsTerm(func=mdp.joint_pos_rel)
        joint_vel_rel = ObsTerm(func=mdp.joint_vel_rel)

        def __post_init__(self) -> None:
            self.enable_corruption = False
            self.concatenate_terms = False

    policy: PolicyCfg = PolicyCfg()

@configclass
class CartpoleEnvCfg(ManagerBasedEnvCfg):
    scene = CartpoleSceneCfg(num_envs=16, env_spacing=2.5)
    observations = ObservationsCfg()
    actions = ActionsCfg()

    def __post_init__(self):
        self.viewer.eye = [4.5, 0.0, 6.0]
        self.viewer.lookat = [0.0, 0.0, 2.0]
        self.decimation = 1
        self.sim.dt = 0.005

def main():
    env_cfg = CartpoleEnvCfg()
    env_cfg.scene.robot.actuators["pole_actuator"].stiffness = 100.0
    env_cfg.scene.robot.actuators["pole_actuator"].damping = 5.0
    env_cfg.sim.physx.solver_type = 1
    env = ManagerBasedEnv(cfg=env_cfg)
    obs, _ = env.reset()
    prev_pos = curr_pos = obs["policy"]["joint_pos_rel"][0][1].item()

    while simulation_app.is_running():
        with torch.inference_mode():
            joint_pos = torch.ones_like(env.action_manager.action) * 5.0
            obs, _ = env.step(joint_pos)
            curr_pos = obs["policy"]["joint_pos_rel"][0][1].item()
            curr_vel = obs["policy"]["joint_vel_rel"][0][1].item()

            dt = env_cfg.sim.dt * env_cfg.decimation
            numerical_vel = (curr_pos - prev_pos) / dt
            print(f"[Env 0]: Pole joint pos: {curr_pos:.8f}, numerical velocity: {numerical_vel:.8f}, measured velocity: {curr_vel:.8f}")
            prev_pos = curr_pos

    env.close()

if __name__ == "__main__":
    main()
    simulation_app.close()

```

**Comment by RandomOakForest** [comment]
Thank you for posting this. We will review and follow up. Can you confirm the versions of Isaac Lab and Isaac Sim you are using? Thanks.

**Comment by ori-gadot** [comment]
Isaac Sim 4.2.0
Isaac Lab 1.3.0

**Comment by ori-gadot** [comment]
Hi @RandomOakForest , I was wondering if there are any updates on this bug report.
I'm also curious about why it was tagged as a question, as it seems to be a bug rather than a question.
Thank you!

**Comment by RandomOakForest** [comment]
Thank you for following up. We are experiencing some delay with the new release and the holidays. The question tag has been removed. We will aim to repro this on our side. Thanks!

**Comment by NitsanBlau** [comment]
Updated code snippet  - Isaaclab 2.3.0/ Issacsim 5.1.0:
```python
from isaaclab.app import AppLauncher

app_launcher = AppLauncher(headless=True)
simulation_app = app_launcher.app

from isaaclab.envs import ManagerBasedEnv
from isaaclab.envs import ManagerBasedEnvCfg
import isaaclab.envs.mdp as mdp
from isaaclab.managers import ObservationGroupCfg as ObsGroup
from isaaclab.managers import ObservationTermCfg as ObsTerm
from isaaclab.utils import configclass
from isaaclab_tasks.manager_based.classic.cartpole.cartpole_env_cfg import CartpoleSceneCfg
import torch
import matplotlib.pyplot as plt


@configclass
class ActionsCfg:
    joint_pos = mdp.JointPositionActionCfg(
        asset_name="robot", joint_names=["cart_to_pole"], scale=1.0, use_default_offset=False
    )


@configclass
class ObservationsCfg:
    @configclass
    class PolicyCfg(ObsGroup):
        joint_pos_rel = ObsTerm(func=mdp.joint_pos_rel)
        joint_vel_rel = ObsTerm(func=mdp.joint_vel_rel)

        def __post_init__(self) -> None:
            self.enable_corruption = False
            self.concatenate_terms = False

    policy: PolicyCfg = PolicyCfg()


@configclass
class CartpoleEnvCfg(ManagerBasedEnvCfg):
    scene = CartpoleSceneCfg(num_envs=16, env_spacing=2.5)
    observations = ObservationsCfg()
    actions = ActionsCfg()

    def __post_init__(self):
        self.viewer.eye = [4.5, 0.0, 6.0]
        self.viewer.lookat = [0.0, 0.0, 2.0]
        self.decimation = 1
        self.sim.dt = 0.005


def main():
    env_cfg = CartpoleEnvCfg()
    env_cfg.scene.robot.actuators["pole_actuator"].stiffness = 100.0
    env_cfg.scene.robot.actuators["pole_actuator"].damping = 5.0
    env_cfg.sim.physx.solver_type = 1
    env = ManagerBasedEnv(cfg=env_cfg)
    obs, _ = env.reset()
    prev_pos = curr_pos = obs["policy"]["joint_pos_rel"][0][1].item()

    # Data aggregation lists
    time_steps = []
    positions = []
    numerical_velocities = []
    measured_velocities = []
    dt = env_cfg.sim.dt * env_cfg.decimation
    step = 0
    max_steps = 1000

    while simulation_app.is_running() and step < max_steps:
        with torch.inference_mode():
            joint_pos = torch.ones_like(env.action_manager.action) * 5.0
            obs, _ = env.step(joint_pos)
            curr_pos = obs["policy"]["joint_pos_rel"][0][1].item()
            curr_vel = obs["policy"]["joint_vel_rel"][0][1].item()

            numerical_vel = (curr_pos - prev_pos) / dt

            # Aggregate data
            time_steps.append(step * dt)
            positions.append(curr_pos)
            numerical_velocities.append(numerical_vel)
            measured_velocities.append(curr_vel)

            prev_pos = curr_pos
            step += 1

    env.close()

    # Create temporal plot
    fig, axes = plt.subplots(3, 1, figsize=(12, 10), sharex=True)

    # Position plot
    axes[0].plot(time_steps, positions, "b-", linewidth=1.5)
    axes[0].set_ylabel("Position (rad)")
    axes[0].set_title("Pole Joint Position Over Time")
    axes[0].grid(True, alpha=0.3)

    # Velocity comparison plot
    axes[1].plot(time_steps, measured_velocities, "g-", label="Measured Velocity", linewidth=1.5)
    axes[1].plot(time_steps, numerical_velocities, "r--", label="Numerical Velocity", linewidth=1.5)
    axes[1].set_ylabel("Velocity (rad/s)")
    axes[1].set_title("Velocity Comparison: Measured vs Numerical")
    axes[1].legend()
    axes[1].grid(True, alpha=0.3)

    # Velocity error plot
    velocity_error = [m - n for m, n in zip(measured_velocities, numerical_velocities)]
    axes[2].plot(time_steps, velocity_error, "m-", linewidth=1.5)
    axes[2].set_ylabel("Error (rad/s)")
    axes[2].set_xlabel("Time (s)")
    axes[2].set_title("Velocity Error (Measured - Numerical)")
    axes[2].grid(True, alpha=0.3)

    plt.tight_layout()
    plt.savefig("velocity_bug_analysis.png", dpi=150)

    # Create zoomed-in plot for second half
    half_idx = len(time_steps) // 2
    time_second_half = time_steps[half_idx:]
    pos_second_half = positions[half_idx:]
    num_vel_second_half = numerical_velocities[half_idx:]
    meas_vel_second_half = measured_velocities[half_idx:]
    error_second_half = velocity_error[half_idx:]

    fig2, axes2 = plt.subplots(3, 1, figsize=(12, 10), sharex=True)

    # Position plot (zoomed)
    axes2[0].plot(time_second_half, pos_second_half, "b-", linewidth=1.5)
    axes2[0].set_ylabel("Position (rad)")
    axes2[0].set_title("Pole Joint Position Over Time (Second Half)")
    axes2[0].grid(True, alpha=0.3)

    # Velocity comparison plot (zoomed)
    axes2[1].plot(time_second_half, meas_vel_second_half, "g-", label="Measured Velocity", linewidth=1.5)
    axes2[1].plot(time_second_half, num_vel_second_half, "r--", label="Numerical Velocity", linewidth=1.5)
    axes2[1].set_ylabel("Velocity (rad/s)")
    axes2[1].set_title("Velocity Comparison: Measured vs Numerical (Second Half)")
    axes2[1].legend()
    axes2[1].grid(True, alpha=0.3)

    # Velocity error plot (zoomed)
    axes2[2].plot(time_second_half, error_second_half, "m-", linewidth=1.5)
    axes2[2].set_ylabel("Error (rad/s)")
    axes2[2].set_xlabel("Time (s)")
    axes2[2].set_title("Velocity Error (Measured - Numerical) (Second Half)")
    axes2[2].grid(True, alpha=0.3)

    plt.tight_layout()
    plt.savefig("velocity_bug_analysis_zoomed.png", dpi=150)
    plt.show()


if __name__ == "__main__":
    main()
    simulation_app.close()

**Comment by michaellin6** [comment]
Hi @NitsanBlau and @ori-gadot, thank you for posting your question. Verifying the correctness of TGS solution with numerically computed velocity is limited to the correctness of the numerical approximation. You are implementing a backward differencing velocity in the script, you may want to try instead a central differencing which is more accurate but you will need to compute that offline. Another approach to verify the correctness of the physics is to use total energy of the system: Kinetic + Potential energy should stay constant.

**Comment by michaellin6** [comment]
It looks like this issue has gone quiet. Since there's been no response for a while, I'll be closing it. If you're still experiencing this problem, please feel free to re-open it with updated information.
